File: cfgen/cfgenlib/kondensor.cfgenlib.policy.elements/spec.py

#
# (c) Copyright 2022, 2023 Kondensor Contributors
# Written by Warwick Molloy.
# Distributed under the Kondensor License.
#

import re

## Convert spec file into C# files.

context = {
  "line_buffer": None,
  "end_of_file": False
}

def getLine():
  line_buffer = context["line_buffer"]
  end_of_file = context["end_of_file"]
  fetch = False
  if line_buffer == None and not end_of_file:
    try:
      line = input()
      fetch = True
    except Exception as e:
      end_of_file = True
      line = None
      print("Ex: ", e)
  else:
    line = line_buffer
    fetch = False
    line_buffer = None
  context["end_of_file"] = end_of_file
  context["line_buffer"] = line_buffer
  return line

def ungetLine(line):
  context["line_buffer"] = line

def is_Eof():
  end_of_file = context["end_of_file"]
  return end_of_file

def getWord(line):
  try:
    if line != None and len(line) == 0:
      return None
    else:
      mm = re.search("([:\w \-\*]+)", line)
      if mm == None:
        ungetLine(line)
        return None
      else:
        return mm.group(1)
  except Exception as e:
    return None

def isDeclarationLine(line):
  if line == None:
    return False
  else:
    tab = line.find("\t")
    return (tab > 1)

def getList():
  list = []
  finished = False
  while not finished and not is_Eof():
    line = getLine()
    if isDeclarationLine(line):
      ungetLine(line)
      finished = True
    else:
      entry = getWord(line)
      if entry != None:
        list.append(entry)
      
  return list

def splitList(list, prefix):
  """
  Split the given list into 2 lists.
  RETURNS (list_no_prefix, list_prefixed)
  The first list returned have items lacking the prefix.
  The second list returned will have only the items with the prefix.
  """
  non_prefixed = []
  prefixed = []
  for item in list:
    if type(item) == str and item.find(prefix) == 0:
      prefixed.append(item)
    else:
      non_prefixed.append(item)
  return (non_prefixed, prefixed)

def ProcessDecl(line):
  #Actions	->|Description	->|Access level	->|Resource types (*required)	->|Condition keys	->|Dependent actions
  mm = re.search("^(\w+)\s([\w \-]+)[\t]?([\w \-]+)[\t]?([\w \-]*)[\t]?", line)
  if mm != None:
    action = mm.group(1)
    description = mm.group(2)
    access = mm.group(3)
    nextLine = getLine()
    resource_types = getWord(nextLine)
    items = getList()
    (conditions, dependent_actions) = splitList(items, "iam")
    print("Action: %s" % action)
    print("Desc: %s" % description)
    print("Access: %s" % access)
    print("Res Types: %s" % resource_types)
    print("%d Conditions: %s" % (len(conditions),conditions) )
    print("%d Dependent Actions: %s" % (len(dependent_actions), dependent_actions))
    print("-----")


def FindContent(line):
  if isDeclarationLine(line):
    ProcessDecl(line)
  else:
    raise Exception("Did not handle line:" + line)

def ReadInput():
  x = ""
  while not is_Eof():
    line = getLine()
    if line != None and len(line) > 0:
      FindContent(line)


ReadInput()

File: policygen/Spec.cs
/*
 *  (c) Copyright 2022, 2023 Kondensor Contributors
 *  Written by Warwick Molloy.
 *  Distributed under the Kondensor License.
 */

using Optional;

using System.Text.RegularExpressions;
using System.IO;


public static class Spec
{
  private static readonly Regex Re = new Regex(
    @"^(\w+)\s([\w \-]+)[\t]?([\w \-]+)[\t]?([\w \-]*)[\t]?",
    RegexOptions.Compiled | RegexOptions.IgnoreCase
  );

  private static readonly Regex WordRe = new Regex(
    pattern: @"([:\w \-\*]+)"
  );

  private static Option<string> OptLineBuffer = Option.None<string>();
  private static bool EndOfFile = false;

  private static void SetEof() => EndOfFile = true;
  private static bool IsEof() => EndOfFile;


  public static Option<string> GetLine(TextReader reader)
  {
    Option<string> line = Option.None<string>();

    if (OptLineBuffer.HasValue)
    {
      OptLineBuffer.MatchSome( text => line = Option.Some(text) );
      OptLineBuffer = Option.None<string>();
    }
    else
    {
      try {
        string? value = reader.ReadLine();
        if (value == null)
        {
          SetEof();
        }
        else
        {
          line = Option.Some(value);
        }
      }
      catch(Exception e)
      {
        Console.WriteLine(
          $"Error encountered when fetching a line of text. {e.Message}"
        );
        SetEof();
      }
    }
    return line;
  }

  public static void UngetLine(Option<string> line)
    => OptLineBuffer = line;

  // def getWord(line):
  // try:
  //   if line != None and len(line) == 0:
  //     return None
  //   else:
  //     mm = re.search("([:\w \-\*]+)", line)
  //     if mm == None:
  //       ungetLine(line)
  //       return None
  //     else:
  //       return mm.group(1)
  // except Exception as e:
  //   return None
  public static Option<string> GetWord(Option<string> line)
  {
    Option<string> wordResult = Option.None<string>();
    line.MatchSome( textLine => {
      if (textLine.Length > 0)
      {
        var match = WordRe.Match(textLine);
        if (match != null && match.Length > 0)
        {
          GroupCollection groups = match.Groups;
          wordResult = Option.Some(groups[1].Value);
        }
        else
        {
          UngetLine(line);
        }
      }
    });
    return wordResult;
  }

  public static void ProcessDeclaration(string line)
  {
    var matches = Re.Matches(line);
    //mm = re.search("^(\w+)\s([\w \-]+)[\t]?([\w \-]+)[\t]?([\w \-]*)[\t]?", line)
    if (matches != null && matches.Count > 0)
    {
      GroupCollection groups = matches[0].Groups;
      Console.WriteLine($"There are {groups.Count} matched groups.");
      var action = groups[1].Value;
      var description = groups[2].Value;
      var access = groups[3].Value;

      // Console.WriteLine($"#0 -> '{groups[0]}'");

      Console.WriteLine($"Action: {action}");
      Console.WriteLine($"Desc: {description}");
      Console.WriteLine($"Access: {access}");
      for (int index = 4; index < groups.Count; index++)
      {
        Console.WriteLine($"#{index} -> '{groups[index]}'");
      }
      //nextLine = getLine()
    // resource_types = getWord(nextLine)
    // items = getList()
    // (conditions, dependent_actions) = splitList(items, "iam")
    // print("Action: %s" % action)
    // print("Desc: %s" % description)
    // print("Access: %s" % access)
    // print("Res Types: %s" % resource_types)
    // print("%d Conditions: %s" % (len(conditions),conditions) )
    // print("%d Dependent Actions: %s" % (len(dependent_actions), dependent_actions))
    // print("-----")

    }
  }

  public static void ReadStream(TextReader reader)
  {
    var line = GetLine(reader);
    line.MatchSome( textLine => ProcessDeclaration(textLine));
    //
  }

}
